marlowe-test/src/Spec/Marlowe/Service/Serialization.hs

```diff
@@ -0,0 +1,106 @@
+-----------------------------------------------------------------------------
+--
+-- Module      :  $Headers
+-- License     :  Apache 2.0
+--
+-- Stability   :  Experimental
+-- Portability :  Portable
+--
+-- | Test of Marlowe's Cardano JSON implementation against the reference implementation.
+--
+-----------------------------------------------------------------------------
+
+
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+
+
+module Spec.Marlowe.Service.Serialization
+  ( -- * Types
+    SerializationResponse(..)
+  , knownJsonTypes
+    -- * Testing
+  , roundtripSerialization
+  ) where
+
+
+import Control.Applicative ((<|>))
+import Data.Aeson (FromJSON(..), ToJSON(..))
+import Data.Jsonable (JsonableType(JsonableType), KnownJsonable, isKnownJson, roundTripJsonable)
+import Data.Proxy (Proxy(..))
+import Spec.Marlowe.Semantics.Arbitrary ()
+
+import qualified Data.Aeson as A (Value, object, withObject, (.:), (.=))
+import qualified Language.Marlowe.Core.V1.Semantics as Marlowe
+import qualified Language.Marlowe.Core.V1.Semantics.Types as Marlowe
+
+
+-- | Response to a round-trip serialization request.
+data SerializationResponse =
+    -- | Success.
+    SerializationSuccess
+    {
+      valueReserialized :: A.Value  -- ^ The reserialized value.
+    }
+    -- | The type was not known.
+  | UnknownType
+    {
+      unknownType :: String  -- ^ The type.
+    }
+    -- | The deserialization or serialization failed.
+  | SerializationError
+    {
+      serializationError :: String  -- ^ The error message.
+    }
+  deriving (Eq, Ord, Read, Show)
+
+instance ToJSON SerializationResponse where
+  toJSON SerializationSuccess{..} = A.object . pure $ "serialization-success" A..= valueReserialized
+  toJSON UnknownType{..} = A.object . pure $ "unknown-type" A..= unknownType
+  toJSON SerializationError{..} = A.object . pure $ "serialization-error" A..= serializationError
+
+instance FromJSON SerializationResponse where
+  parseJSON =
+    A.withObject "SerializationResponse"
+      $ \o ->
+            (SerializationSuccess <$> o A..: "serialization-success")
+        <|> (UnknownType <$> o A..: "unknown-type")
+        <|> (SerializationError <$> o A..: "serialization-error")
+
+
+-- | Deserialize and then serialize a value.
+roundtripSerialization
+  :: String  -- ^ The key to the type.
+  -> A.Value  -- ^ The value.
+  -> SerializationResponse  -- ^ The result.
+roundtripSerialization typeSerialized valueSerialized =
+  if isKnownJson knownJsonTypes typeSerialized
+    then case roundTripJsonable knownJsonTypes typeSerialized valueSerialized of
+             Right valueReserialized -> SerializationSuccess{..}
+             Left serializationError      -> SerializationError{..}
```

Comment:
```suggestion
             Left serializationError  -> SerializationError{..}
```
Just strange indentation

Reply:
Fixed in c4cbff6a21bc2da875aa867375658d8f8fe803a6.